File: runtime/dart/convert/chunkedConversionSink.ts
import { IList } from "../core/list";
import { ISink } from "../core/sink";
declare const dart: {
    convert: {
        chunkedConversionSink: <T>(
            this: void,
            chunkedConversionSink: IChunkedConversionSink<T>
        ) => IChunkedConversionSink<T>;
        chunkedConversionSinkWithCallback: <T>(
            callback: (accumulated: IList<T>) => void
        ) => IChunkedConversionSink<T>;
    };
};
export interface IChunkedConversionSink<T> {
    add: (chunk: T) => void;
    close: () => void;
    toString: () => string;
    getHashCode: () => number;
}
export class ChunkedConversionSink<T> implements ISink<T> {
    public constructor() {
        dart.convert.chunkedConversionSink(this);
    }
    public static withCallback<T>(
        callback: (accumulated: IList<T>) => void
    ): IChunkedConversionSink<T> {
        return dart.convert.chunkedConversionSinkWithCallback(callback);
    }
    private readonly _dart_add: (chunk: T) => void = undefined as any;
    private readonly _dart_close: () => void = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    public add(chunk: T): void {
        return this._dart_add(chunk);
    }
    public close(): void {
        return this._dart_close();
    }
    public toString(): string {
        return this._dart_toString();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
}

File: lib/cfr/builtins/libs/dart/convert/chunkedConversionSink.dart
import 'dart:convert';
import 'dart:core';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedChunkedConversionSink
    extends VMManagedBox<ChunkedConversionSink<dynamic>> {
  VMManagedChunkedConversionSink(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['add'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.add(luaCallerArguments[1]);
      return [];
    });
    table['close'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.close();
      return [];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final ChunkedConversionSink vmObject;
}

class RTManagedChunkedConversionSink extends ChunkedConversionSink
    implements Box<ChunkedConversionSink> {
  RTManagedChunkedConversionSink(
      {required this.table, required this.hydroState})
      : super() {
    table['vmObject'] = vmObject;
    table['unwrap'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [unwrap()];
    });
    table['_dart_add'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      add(luaCallerArguments[1]);
      return [];
    });
    table['_dart_close'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      close();
      return [];
    });
    table['_dart_toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.toString()];
    });
    table['_dart_getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.hashCode];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  ChunkedConversionSink unwrap() => this;
  ChunkedConversionSink get vmObject => this;
  @override
  void add(dynamic chunk) {
    Closure closure = table["add"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  void close() {
    Closure closure = table["close"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  String toString() {
    Closure closure = table["__tostring"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  int get hashCode {
    Closure closure = table["getHashCode"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }
}

void loadChunkedConversionSink(
    {required HydroState hydroState, required HydroTable table}) {
  table['chunkedConversionSink'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      RTManagedChunkedConversionSink(
          table: luaCallerArguments[0], hydroState: hydroState)
    ];
  });
  table['chunkedConversionSinkWithCallback'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    Closure unpackedcallback = luaCallerArguments[1];
    return [
      maybeBoxObject<ChunkedConversionSink>(
          object: ChunkedConversionSink.withCallback(
              (accumulated) => unpackedcallback.dispatch(
                    [luaCallerArguments[0], accumulated],
                    parentState: hydroState,
                  )),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<ChunkedConversionSink>(boxer: (
      {required ChunkedConversionSink vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedChunkedConversionSink(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}
